### ISSUE TITLE
Date arithmetic works when nopython=False, but not when nopython=True

### ISSUE THREAD

**Issue by cpcloud** [numba/numba#1202]
`nopython=False`:

``` python
In [14]: from numba import jit, vectorize

In [15]: f = jit(NPDatetime('D')(NPDatetime('D')))(lambda x: x + np.timedelta64(1, 'D'))

In [16]: x = np.random.randint(10, size=10).astype('datetime64[D]')

In [17]: f(x)
Out[17]:
array(['1970-01-10', '1970-01-04', '1970-01-07', '1970-01-03',
       '1970-01-03', '1970-01-10', '1970-01-04', '1970-01-07',
       '1970-01-03', '1970-01-02'], dtype='datetime64[D]')
```

`nopython=True`:

```
UntypedAttributeError: Caused By:
Traceback (most recent call last):
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/compiler.py", line 201, in run
    res = stage()
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/compiler.py", line 415, in stage_nopython_frontend
    self.locals)
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/compiler.py", line 710, in type_inference_stage
    infer.propagate()
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/typeinfer.py", line 408, in propagate
    self.constrains.propagate(self.context, self.typevars)
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/typeinfer.py", line 107, in propagate
    constrain(context, typevars)
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/typeinfer.py", line 275, in __call__
    attrty = context.resolve_getattr(value=ty, attr=self.attr)
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/typing/context.py", line 88, in resolve_getattr
    ret = attrinfo.resolve(value, attr)
  File "/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numba/typing/templates.py", line 335, in resolve
    raise UntypedAttributeError(value=value, attr=attr)
UntypedAttributeError: Unknown attribute "timedelta64" of type Module(<module 'numpy' from '/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numpy/__init__.pyc'>)

Failed at nopython (nopython frontend)
Unknown attribute "timedelta64" of type Module(<module 'numpy' from '/Users/pcloud/miniconda/envs/py27/lib/python2.7/site-packages/numpy/__init__.pyc'>)
```

**Comment by pitrou** [comment]
Ideally it would work better this way:

```
one_day = np.timedelta64(1, 'D')

@jit(nopython=True)
def f(x):
    return x + one_day
```

but there's still an error which I'm going to try and solve.

**Comment by pitrou** [comment]
Oh, I'm seeing you're doing array arithmetic with datetime64 / timedelta64. That's not supported yet, because we rely on the standard Numpy ufuncs type information, and that type information doesn't take datetime units into account (e.g. 'D').

**Comment by aseem-hegshetye** [comment]
Being able to use datetime64 arrays in numba would be great.

**Comment by dscole** [comment]
What is the status of this issue? I'm also trying to do datetime/timedelta arithmetic which doesn't currently work in numba (further discussion in #917). Right now I use the following workaround (based on https://github.com/numba/numba/issues/917#issuecomment-68448844) and I was wondering can't this be implemented more generically inside numba.
I want to be able to write:
```
 @nb.jit(nopython=True)
def foo(a, b):
    return a + b
```
But that doesn't work when `a` and `b` are `array(timedelta64(ns), 1d, C)`, but the following does:
```
 @nb.jit(nopython=True)
def foo(a, b):
    original = a.dtype
    return (a.view(nb.int64) + b.view(nb.int64)).view(original)

a = np.array([1], np.dtype('m8[ns]'))
b = np.array([1], np.dtype('m8[ns]'))
foo(a, b)
```
In other words add support for ufuncs for datetime/timedeltas that use the same time units, by simply treating them as int64. Just having that ability would actually help my use case a lot. But even operating on different time units shouldn't be that hard either, simply adding a conversion, no?
